File: arena-xunit/tests/Unit/ContainerizedComponentBuilderSerializationTest.cs
using ArenaDotnet.Xunit.Component;
using Newtonsoft.Json.Linq;
using Xunit;

namespace ArenaDotnet.Xunit.UnitTest;

public class ContainerizedComponentBuilderSerializationTest
{
    [Fact]
    public void Build_WithContainerfile_SerializesCorrectJson()
    {
        var comp = new ContainerizedComponentBuilder("test").WithContainerfile("./Dockerfile").Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("container", obj["type"]);
        Assert.Equal("./Dockerfile", obj["containerfile"]);
        Assert.NotNull(obj["identifier"]);
    }

    [Fact]
    public void Build_WithEnvVar_SerializesCorrectJson()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithEnvVar("KEY", "value")
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        var envVars = obj["env_vars"];
        Assert.NotNull(envVars);
        Assert.Equal("value", envVars["KEY"]);
    }

    [Fact]
    public void Build_WithRuntimeArg_SerializesCorrectJson()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithRuntimeArg("flag", "arg1")
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        var runtimeArgs = obj["runtime_args"];
        Assert.NotNull(runtimeArgs);
        var arg = Assert.Single(runtimeArgs);
        Assert.Equal("flag", arg["name"]);
        Assert.Equal("arg1", arg["value"]);
    }

    [Fact]
    public void Build_WithBuildContextImageTagAndNetwork_SerializesCorrectJson()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithBuildContext("./ctx")
            .WithImageTag("v1")
            .WithNetwork("test-net")
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("./ctx", obj["build_context"]);
        Assert.Equal("v1", obj["image_tag"]);
        Assert.Equal("test-net", obj["network"]);
    }

    [Fact]
    public void Build_WithPortMapping_SerializesHostAndContainerPort()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithPortMapping(8080, 80)
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        var portMappings = obj["port_mappings"];
        Assert.NotNull(portMappings);
        var mapping = Assert.Single(portMappings);
        Assert.Equal(8080, mapping["host_port"]);
        Assert.Equal(80, mapping["container_port"]);
    }

    [Fact]
    public void Build_WithHostMapping_SerializesCorrectJson()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithHostMapping("host.docker.internal:host-gateway")
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        var hostMappings = obj["host_mappings"];
        Assert.NotNull(hostMappings);
        var mapping = Assert.Single(hostMappings);
        Assert.Equal("host.docker.internal:host-gateway", mapping);
    }

    [Fact]
    public void Build_WithVolumeMapping_SerializesHostAndContainerPath()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithVolumeMapping("/host/path", "/container/path")
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        var volumeMappings = obj["volume_mappings"];
        Assert.NotNull(volumeMappings);
        var mapping = Assert.Single(volumeMappings);
        Assert.Equal("/host/path", mapping["host_path"]);
        Assert.Equal("/container/path", mapping["container_path"]);
    }

    [Fact]
    public void Build_WithReadinessCheck_SerializesHttpKind()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithReadinessCheck(HttpReadinessCheck.Create(), "http://127.0.0.1:8080/health")
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        var readinessChecks = obj["readiness_checks"];
        Assert.NotNull(readinessChecks);
        var check = Assert.Single(readinessChecks);
        Assert.Equal("http", check["kind"]);
    }

    [Fact]
    public void Build_WithoutContainerfile_Throws()
    {
        var builder = new ContainerizedComponentBuilder("test");
        Assert.Throws<System.InvalidOperationException>(() => builder.Build());
    }

    [Fact]
    public void Build_WithContainerfileAndImage_Throws()
    {
        var builder = ContainerizedComponentBuilder.FromImage("test", "postgres:18-bookworm")
            .WithContainerfile("./Dockerfile");
        Assert.Throws<System.InvalidOperationException>(() => builder.Build());
    }

    [Fact]
    public void Build_FromImage_SerializesImageWithoutContainerfile()
    {
        var comp = ContainerizedComponentBuilder.FromImage("test", "postgres:18-bookworm").Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("container", obj["type"]);
        Assert.Equal("postgres:18-bookworm", obj["image"]);
        Assert.Null(obj["containerfile"]);
    }

    [Fact]
    public void Build_WithPlatform_SerializesPlatform()
    {
        var comp = new ContainerizedComponentBuilder("test")
            .WithContainerfile("./Dockerfile")
            .WithPlatform("linux/arm64")
            .Build();
        var json = comp.ForFfi();
        var obj = JObject.Parse(json);
        Assert.Equal("linux/arm64", obj["platform"]);
    }

    [Fact]
    public void Build_Identifier_MatchesPattern()
    {
        var comp = new ContainerizedComponentBuilder("test").WithContainerfile("./Dockerfile").Build();
        Assert.StartsWith("arena-container-", comp.Identifier);
    }
}

File: arena-junit/src/test/java/arena/junit/exec/ContainerizedComponentBuilderSerializationTest.java
package arena.junit.exec;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import arena.junit.match.ArenaRunnableComponent;
import arena.junit.readiness.HttpReadinessCheck;
import arena.junit.readiness.TcpReadinessCheck;
import arena.junit.support.ArenaJson;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import org.junit.jupiter.api.Test;

final class ContainerizedComponentBuilderSerializationTest {

  static final class StubComponent implements ArenaRunnableComponent {
    @Override
    public ObjectNode forFfi() {
      return ArenaJson.object().put("identifier", "child");
    }
  }

  @Test
  void build_minimalNameAndContainerfile_serializesTypeAndEmptyCollections() {
    ObjectNode config = new ContainerizedComponentBuilder("web", "Containerfile").build().forFfi();
    assertEquals("container", config.path("type").asText());
    assertTrue(config.path("identifier").asText().startsWith("arena-containerized-component-web-"));
    assertEquals("Containerfile", config.path("containerfile").asText());
    assertTrue(config.path("env_vars").isEmpty());
    assertTrue(config.path("runtime_args").isEmpty());
    assertTrue(config.path("port_mappings").isEmpty());
    assertTrue(config.path("host_mappings").isEmpty());
    assertFalse(config.has("readiness_checks"));
    assertFalse(config.has("children"));
  }

  @Test
  void withBuildContextNetworkAndImageTag_setsScalarFields() {
    ObjectNode config =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .withBuildContext(".")
            .withNetwork("net")
            .withImageTag("v1")
            .build()
            .forFfi();
    assertEquals(".", config.path("build_context").asText());
    assertEquals("net", config.path("network").asText());
    assertEquals("v1", config.path("image_tag").asText());
  }

  @Test
  void withPortMapping_appendsHostAndContainerPorts() {
    ObjectNode config =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .withPortMapping(8080, 80)
            .build()
            .forFfi();
    ObjectNode mapping = (ObjectNode) config.path("port_mappings").get(0);
    assertEquals(8080, mapping.path("host_port").asInt());
    assertEquals(80, mapping.path("container_port").asInt());
  }

  @Test
  void withHostMapping_appendsHostMappingsArray() {
    ObjectNode config =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .withHostMapping("db.local:127.0.0.1")
            .build()
            .forFfi();
    assertEquals(
        "db.local:127.0.0.1", config.path("host_mappings").get(0).asText());
  }

  @Test
  void withEnvVar_setsEnvVarsObject() {
    ObjectNode config =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .withEnvVar("LOG_LEVEL", "debug")
            .build()
            .forFfi();
    assertEquals("debug", config.path("env_vars").path("LOG_LEVEL").asText());
  }

  @Test
  void withRuntimeArg_appendsNameValuePair() {
    ObjectNode config =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .withRuntimeArg("--cap-add", "NET_ADMIN")
            .build()
            .forFfi();
    ObjectNode pair = (ObjectNode) config.path("runtime_args").get(0);
    assertEquals("--cap-add", pair.path("name").asText());
    assertEquals("NET_ADMIN", pair.path("value").asText());
  }

  @Test
  void withReadinessCheck_defaultTimeout_usesTenSecondDefault() {
    ContainerizedComponent component =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .withReadinessCheck(HttpReadinessCheck.create(), "/health")
            .build();
    assertEquals(10_000L, component.readinessEntries().getFirst().timeoutMs());
  }

  @Test
  void withReadinessCheck_httpAndTcp_serializeReadinessChecksArray() {
    ObjectNode config =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .withReadinessCheck(HttpReadinessCheck.create(), "/health", 5_000L)
            .withReadinessCheck(TcpReadinessCheck.create(), "5432", 2_000L)
            .build()
            .forFfi();
    ArrayNode checks = (ArrayNode) config.path("readiness_checks");
    assertEquals("http", checks.get(0).path("kind").asText());
    assertEquals("/health", checks.get(0).path("target").asText());
    assertEquals(5_000L, checks.get(0).path("timeout_ms").asLong());
    assertEquals("tcp", checks.get(1).path("kind").asText());
    assertEquals("5432", checks.get(1).path("target").asText());
  }

  @Test
  void addChildComponent_nonEmptyChildren_serializesChildrenArray() {
    ObjectNode config =
        new ContainerizedComponentBuilder("web", "Containerfile")
            .addChildComponent(new StubComponent())
            .build()
            .forFfi();
    assertEquals(1, ((ArrayNode) config.path("children")).size());
  }

  @Test
  void identifier_returnsConfiguredIdentifier() {
    ContainerizedComponent component =
        new ContainerizedComponentBuilder("web", "Containerfile").build();
    assertTrue(component.identifier().startsWith("arena-containerized-component-web-"));
  }
}
